dynamiqs/mesolve/mesolve.py

```diff
@@ -101,9 +101,9 @@ def mesolve(
     # === convert arguments
     H = _astimearray(H)
     jump_ops = [_astimearray(L) for L in jump_ops]
-    rho0 = jnp.asarray(rho0, dtype=cdtype())
+    rho0 = _asqarray(rho0)
```

Comment:
What happens if the user passes a `QArray` as the Hamiltonian, but an `Array` as the initial state? We need to have both of the same type, otherwise there will be type promotion during the code and diffrax will crash.

Hence why I'd suggest promoting systematically to a `QArray`. Also, it makes the output typing of `sesolve`/`mesolve` simpler.

Reply:
Note: I thought initially that we could have rules within `QArray` matrix products, e.g. `QArray @ Array -> Array`, but this seems incompatible with the right-hand-side multiplication that will be called in the jump term of the ME, i.e. we'd like `Array @ QArray -> Array` (to handle right-hand-side ME) but also `Array @ QArray -> QArray` (to handle initial state as `QArray`).

Reply:
As discussed on Slack, would it work to just promote everyone to a `QArray` if any argument is a `QArray`?

Reply:
Yes, but it's somewhat more annoying since we have to check all arguments together, instead of checking them one by one indendently. It's technically difficult, just a bit less clean IMO.

Reply:
I guess this is solved by the discussion on the array API 👌